### ISSUE TITLE
Receiving unknown OBJECT messages (who creates the track-id)

### ISSUE THREAD

**Issue by kixelated** [moq-wg/moq-transport#145]
#141 adds a `Track ID` field. This is used to compress a long string (in SUBSCRIBE) into a short varint (in OBJECT).

The problem is that the mapping from string to varint is indicated in the SUBSCRIBE_OK message, which is sent in parallel with OBJECT messages. This introduces a race condition, where packet loss will cause SUBSCRIBE_OK to arrive late. The receiver will be unable to associate OBJECT messages with tracks until the arrival of the SUBSCRIBE_OK.

A correct implementation must buffer any OBJECT messages with unknown track IDs for an indefinite amount of time. Any other implementation will error or skip media due to decode errors. This also introduce a risk of resource exhaustion, for example if the stream containing SUBSCRIBE_OK is reset.

**Comment by kixelated** [comment]
My proposal is to move the `Track ID` to the SUBSCRIBE message. The OBJECT sender MUST use a `Track ID` in a received SUBSCRIBE message. This means the receiver will never receive an unknown or unrequested OBJECT.

**Comment by kixelated** [comment]
I know @wilaw wants the sender to be able to choose the `Track ID`, but that makes this race condition unavoidable. The only way to prevent it is to operate like QPACK; a literal track name is used until the SUBSCRIBE_OK is somehow acknowledged.

The argument is that the sender should choose the Track ID so it could use the same ID across multiple connections. There's a claim that this would improve edge efficiency, but I am extremely doubtful. My time spent profiling QUIC stacks makes me believe that per-hop properties (ex. QUIC connection ID, QUIC steam ID, HTTP/3 QPACK, WebTransport session ID) contribute an insignificant amount to CPU usage.

**Comment by suhasHere** [comment]
At a high level, all the nodes needs to be able to handle unknown objects - its true regardless of trackId issue.

A typical way is to we define the following in the text
- send full track-names + trackId multiple times for the initial X time units over a separate message
- drop objects after Y time units 

We need some text on handling such cases as recommendations from security perspective too.

**Comment by afrind** [comment]
As an Individual:

In many ways this ID is to moq what the QUIC stream ID is to HTTP/3 - and for those, there is a simple, unilateral, and deterministic way to choose.  Maybe that is a good model to follow?  Do we want to allow re-use of Track IDs for different tracks over the life of a connection?

 > A correct implementation must buffer any OBJECT messages with unknown track IDs for an indefinite amount of time. 
 
There's an alternative, which is that the sender uses the subscriber's track ID until the client sends an explicit message acknowledging the new ID (NEW_OBJECT_ID_OK).  Then the server switches to the new ID, and the client has to accept OBJECTs with either ID for some period of time, but it would never have to accept unknown IDs.

**Comment by kixelated** [comment]
> At a high level, all the nodes needs to be able to handle unknown objects - its true regardless of trackId issue.

I don't understand what you mean. The receiver should always be able to identify which the track each OBJECT belongs to, and it should have agreed to receive that track in the first place.

> - send full track-names + trackId multiple times for the initial X time units after subscribe_ok, over a separate message or a variation of the object message

@afrind's suggestion is more fault-tolerant than a timer. 

> - drop unknown objects after Y time units (this is needed independent of track id concerns)

While receivers need the ability to drop OBJECTs in the first place, I don't want to degrade the user experience further. This is a solvable problem.

**Comment by kixelated** [comment]
> In many ways this ID is to moq what the QUIC stream ID is to HTTP/3 - and for those, there is a simple, unilateral, and deterministic way to choose. Maybe that is a good model to follow? Do we want to allow re-use of Track IDs for different tracks over the life of a connection?

Yeah, I would like each SUBSCRIBE message to contain an incrementing `Subscription ID`. The OBJECT, SUBSCRIBE_OK, SUBSCRIBE_ERROR, and UNSUBSCRIBE messages would reference the `Subscription ID`, not the full track name. The ability to reuse subscription IDs seems pointless and potentially problematic (much like QUIC stream IDs).

> There's an alternative, which is that the sender uses the subscriber's track ID until the client sends an explicit message acknowledging the new ID (NEW_OBJECT_ID_OK). Then the server switches to the new ID, and the client has to accept OBJECTs with either ID for some period of time, but it would never have to accept unknown IDs.

Here's what that would look like:
```
-> SUBSCRIBE 1 foo/bar
<- SUBSCRIBE_OK 1 = 5
<- OBJECT 1
-> SUBSCRIBE_OK_ACK 1 = 5
<- OBJECT 5 
```

But it won't quite work because both endpoints could simultaneously decide to use the same ID.

```
-> SUBSCRIBE 1 foo/bar
-> SUBSCRIBE 5 oh/no   <- SUBSCRIBE_OK 1 = 5
```

To fix this, we would have to shard the IDs, ie. sender uses even IDs. Alternatively, the receiver could refuse to send the SUBSCRIBE_OK_ACK when a collision is detected.

Although frankly, I'm not sure why the receiver would ever send a SUBSCRIBE_OK_ACK. It's an optional message that just makes things more complicated for the receiver, necessitating multiple hashmap keys.

**Comment by kixelated** [comment]
While these suggestions are valid, I don't understand the benefit of letting the sender choose the track ID rather than the receiver. It's a near zero performance improvement that is immediately cannibalized by the extra complexity.

**Comment by wilaw** [comment]
> I know @wilaw wants the sender to be able to choose the Track ID, but that makes this race condition unavoidable. 

@wilaw has thought about this and now agrees that the advantage of a receiver choosing the trackID is to dispense with the SUBSCRIBE_OK_ACK and make for a simpler subscribe handshake. It also removes the requirement for the receiver to buffer unknown objects, as the OP has pointed out.  For efficiency,  I think I could implement a hashing scheme to avoid hierarchical nodes changing trackIDs at every hop. I know others think this may not be a performance gain, but I'd still like some way to implement it if I choose. 

Option A: receiver hinting, sender authoritative


> -> SUBSCRIBE 1 foo/bar // receiver indicating that 1 should be used
> <- SUBSCRIBE_OK 1  = 5 // sender responds nope, I'd like to use 5
> <- OBJECT 5 // sender uses new ID right away
> -> SUBSCRIBE 5 oh/no   // receiver asking if 5 is available
> -> SUBSCRIBE_OK_ACK 1 = 5
> <- SUBSCRIBE_OK 5 = 6 // sender responds nope, I'd like to use 6 because I know I have already allocated 5. 
> -> SUBSCRIBE_OK_ACK 5 = 6
> <- OBJECT 6



Option B: receiver authoritative

> -> SUBSCRIBE 1 foo/bar // receiver indicating that 1 must be used
> <- SUBSCRIBE_OK 1  // sender responds yes, I'll do that. 
> <- OBJECT 1 // sender uses 1 
> -> SUBSCRIBE 5 oh/no   // receiver indicating that 5 must be used
> <- SUBSCRIBE_OK 5  // sender responds yes, I'll do that. 
> <- OBJECT 5 // sender uses 5

Of the two, I prefer B.

If congestion causes the SUBSCRIBE_OK to arrive after the object, then no harm because the receiver will know how to interpret the trackID since it asked for it. This raises three corollaries that we need to deal with:
1. Do we even need the SUBSCRIBE_OK message? It's not adding value other than giving the receiver some confidence that its subscribe message was received. Perhaps we could assume it has been received and approved, unless a SUBSCRIBE_REJECTED or other type of message is sent? 
2. What should the receiver do with an object with a trackID that it never asked for? Should it send a UNKNOWN_TRACKID message to the sender to inform it that it is using an unrequested ID?
3. What should the sender do if it is asked to use a trackID that is already in use? I'd suggest it go ahead and use it and let the receiver, which is the entity making the mistake, deal with the conflict.

**Comment by kixelated** [comment]
> For efficiency, I think I could implement a hashing scheme to avoid hierarchical nodes changing trackIDs at every hop. I know others think this may not be a performance gain, but I'd still like some way to implement it if I choose.

That's a cool idea, and a good reason why the receiver should be able to choose a subscription ID instead of it incrementing by one. Within the CDN network: `subscription.id = hash_u64(track.name)`. You'll still need some code to deal with collisions.

> 1. Do we even need the SUBSCRIBE_OK message? It's not adding value other than giving the receiver some confidence that its subscribe message was received. Perhaps we could assume it has been received and approved, unless a SUBSCRIBE_REJECTED or other type of message is sent?

Not at the moment, but I think we might want to throw some more information in there. I think the starting group ID could be useful.


> 2. What should the receiver do with an object with a trackID that it never asked for? Should it send a UNKNOWN_TRACKID message to the sender to inform it that it is using an unrequested ID?

IMO close the connection with an error code. 


> 3. What should the sender do if it is asked to use a trackID that is already in use? I'd suggest it go ahead and use it and let the receiver, which is the entity making the mistake, deal with the conflict.

IMO close the connection with an error code.

**Comment by fluffy** [comment]
I think this bug would better be named "what creates the track-id".  

Let me outline the viable possibilities I see:

- Option A: client picks the track-id and it is passed in subscribe or pub-interest to server 

- Option B: client suggests track-id,  and server picks the track-id. Client waits to get track-id before sending any objects 

- Option C:  client suggests,  and server picks track-id. Object send before the track-id is known to client are sent with full track-name. This would mean subscribed needs a 3 way handshake.

**Comment by fluffy** [comment]
On the pro's / con's of the three options we have:

1. Option A and C allow the media to not be delayed 1 RTT to start flowing. Note that 1 RTT to the CDN edge may not actually be a very big deal given how many RTT to set up the media in the first place. But clearly 0-RTT is a an awesome goal. 

2. Option B and C let the relay choose. There are a bunch of reason for this but to me, one of the high order bits is we are trying to slot into small modification to existing CDN system and this allows the CDN to use these bits to map to identifiers they already have. For CDN's that want the bits to be unique across the CDN, they can do that. For CDN that don't care, they can use whatever the client suggested and not have to deal with the hassle of sending track-names as suggested in option C. 

The extra hassle of option C over A seems worth it to me to give CDNs the flexibility. For CDNs that don't want to use this,  well they can just implement Option A. For clients, I don't see much extra work between A and C.

**Comment by fluffy** [comment]
Trying to summarize my somewhat crappy recollection of old conversations on this. 

On some CDNs, it is very desirable to have a unique identifier that is the CDN can use for deciding how to forward a given object or chunk of bytes from that object. On other CDNs it might not matter.  In some cases this can make it easier to map on to existing HTTP CDNs. 

This argues for staying with the current solution of server picks the ID.  This raises the question of deal with what happens with receiving an object before the message that has the ID.   

A variety solutions have been discussed for this. 

1. buffer received objects until the ID is received - no one loves this but you might have to do it for other reasons anyway 
2. three way handshake so the object is not sent before ID 
3. send ID on same QUIC stream as first object in the group 
4. look at other ways to solve this problem of media before ID 

The client people point out, the ID could be shorter if they are unique only across the stream not unique across multiple clients and that the relay can just keep a mapping table of trackID and session flow to whatever ID they wish was the trackID. However some of the relay side people argue that is hard to make work with HA as the HA solution need the client to pass the relay the ID they can use to lookup and recover the HA state and track ID is perfect for that. 

Another solution that was discussed was the client proposes a ID, if it is OK with the relay it uses it, if the relay is not OK with the proposed ID, it rejects the SUBSCRIBE with an error and provided an alternative ID that the client uses in a second SUBSCRIBE.  There is a PR for this somewhere. It may be a good middle ground.

**Comment by kixelated** [comment]
Okay, talking to Alan I realized that this issue is far more widespread than I realized and doesn't require any packet loss.

Let's assume the application writes the SUBSCRIBE_OK on stream 5. The QUIC library does not send a packet yet, either due to pacing, congestion control, or something like Nagle's algorithm. The application then writes the first OBJECT on stream 31.

The problem is that STREAM frames can be coalesced into the same QUIC packet in an undefined order, and when received will be flushed to the application in an undefined order. Even a QUIC sender encodes in priority order (not required), it depends on the QUIC receiver emitting to the application in priority order, which is possible with a callback-based API but not a wakeup-based API (ie. multi-threaded). The SUBSCRIBE_OK is tiny so the STREAM frame will almost always be coalesced with other STREAM frames.

And this is assuming the control stream is higher priority than the data streams. If that's not done, then it's extremely likely that the N data streams will arrive before the 1 control stream, especially because the control stream will be at the bottom of the round-robin stack.

**Comment by suhasHere** [comment]
Couple of observations 
1. Control streams must always be sent at a higher priority
2.  I am inclined to work towards a implementation that supports the proposal from Cullen here :https://mailarchive.ietf.org/arch/msg/moq/bRFK0DQB3XBBaYGD_GsOJNZ9neg/ ( as referred to above https://github.com/moq-wg/moq-transport/issues/145#issuecomment-1648778532)

This will give us client picking the TrackId and also allowing the flexibility where the server can choose to TrackId.

**Comment by vasilvv** [comment]
"A more flexible protocol is more complex to specify and implement; variations, especially those that are not commonly used, can create potential interoperability hazards. In the absence of strong reasons to be flexible, a simpler protocol is more likely to successfully interoperate." ([RFC 9413](https://www.rfc-editor.org/rfc/rfc9413.html#name-flexible-protocols))

We'd need a strong reason to support both.  So far I've heard a lot of good arguments in favor of the subscriber picking the ID, mostly because this means that the receiver doesn't have to deal with unknown IDs.  I seem to recall one argument in favor of the other approach (server wanting to store state in the subscription ID for incoming objects), but I believe we did resolve it (server acts as a subscriber in those contexts, so it can pick the ID under "subscriber picks" design).  If there are other arguments I am forgetting, please document them here.

**Comment by kixelated** [comment]
> 1. Control streams must always be sent at a higher priority

Unfortunately, this doesn't matter much, since QUIC can coalesce multiple STREAM frames into a single packet. The small  SUBSCRIBE_OK will likely be received in the same packet as the first chunk of the OBJECT. In fact, even with multiple packets, QUIC libraries will use something like `recvmmsg` to receive and process multiple packets at once. The receiving QUIC library will then flush these to the application in an undefined order. 

It's implementation dependent, but this will almost certainly result in a race. A single-threaded library like [Quiche](https://docs.rs/quiche/latest/quiche/struct.Connection.html#method.readable) might iterate based on stream ID, last received frame, or even random order. A multi-threaded library will almost certainly race, as they work by issuing a thread wakeup to the application thread blocked on the stream read.

So even with correct prioritization and zero packet loss, this race will occur on most QUIC implementations.

> This will give us client picking the TrackId and also allowing the flexibility where the server can choose to TrackId.

Can you or Cullen elaborate more what you mean by "flexibility"? It sounds like this is a performance concern, but I don't think it's been clearly stated why. In SF, Cullen briefly mentioned that this is a hardware concern, but a written description of the issue would help.

Note that I'm proposing that the **_receiver_** chooses the `track_id`, regardless of if they are a server or client. You could still use the same `track_id` within your network if relays are deterministic. However, I'm still skeptical of the performance argument sans multicast QUIC.

**Comment by fluffy** [comment]
Just a note for myself that I don't think we have come to consensus in WG on this yet.

**Comment by ianswett** [comment]
Interim notes: There was consensus that the current draft has an issue that should be fixed.

Previously a compromise solution was to reject if the server wants to pick and then the client subscribes with that ID.

Christian will write up a proposal for using two IDs (client and server) in the initial Objects until the subscriber knows the ID.

**Comment by afrind** [comment]
Individual comment: 

I reviewed #262, but wanted to keep this longer comment on the issue instead of the PR.  I may have rambled.  Bear with me.

I have implemented a relay where each subscriber may be using a different track ID.  It has a little loop when forwarding an object, like this:

```
for each moq_session, out_track_id in subscribers:
  # override track_id in the object header with the per-subscriber one
  moq_session.publish(out_track_id, obj_header, payload)
```

In Christian's proposal, since not all subscribers will have confirmed the publisher chosen track id when an upstream object arrives and needs forwarding, I'm not sure the loop changes meaningfully.  I suppose I can have two loops, one over subscribers that are using the upstream track_id for this subscription and one not?  But I have a hard time seeing the win:

```
for each moq_session in confirmed_subscribers:
  # obj_header has the right track ID already, no override, yay!
  moq_session.publish(obj_header, payload)

for each moq_session, out_track_id in unconfirmed_subscribers:
  # these subscribers haven't confirmed using server chosen yet, pass the override
  moq_session.publish(out_track_id, obj_header, payload)
```

What about cached objects?

When the object is coming from cache, there may not be an upstream track ID anymore.  I'm not sure what track ID the relay would propose in Subscribe Ok in this case - does it make sense to cache the upstream track ID?  It's possible different objects in the same track were retrieved over different subscriptions (failover) so I think not.  So the 'serve from cache' path also needs to use the different track ID per subscriber path.  

Upstream Track ID Collisions

There is also the case where the track ID being used by the relay's upstream for a given track is already taken on a particular downstream session.  This can happen when the relay routes subscriptions to two or more different upstreams:

```
S1 -- track "A" id = 1 -\ 
                          R ------- C
S2 -- track "B" id = 1 -/ 
```

One of these tracks also needs a per-subscriber track ID, even if we make no changes to the draft.  Maybe this is why hashing the track name is mentioned above -- that solves some things, though it would seem like it's sacrificing some of the compressive benefit of these IDs, and compressing a full-track-name is their primary function.

Clients publishing via Relays

I believe clients and servers need to have independent number spaces for proposed vs publisher chosen track IDs (eg odd/even) to avoid a different type of race condition (see my comment on the PR).  Maybe I'm wrong, but if not this would mean that `C1 -- R -- C2` can never reuse a track ID, since C1 -> R would publish on track IDs with the opposite parity from R -> C2.

Can we estimate the benefits?

The proponents of publisher chosen IDs have cited both memory and performance as potential reasons for requiring this (please add other reasons if I missed them).

For memory it can save 8 bytes per subscription with confirmed ID.  Is there any back-of-the-envelope math indicating this adds up to something significant?  What's the target number of active subscriptions per-server (1M, 10M, 100M, 1B)?
  
For performance, we can write a benchmark comparing the cost writing the same object header +  payload vs unique object headers per payload for say, 100,000 subscribers.  What's the target for number of objects forwarded per-second per-server?

If we have rough numbers showing we cannot meet performance goals that would be a compelling reason to add the SUBSCRIBE CONFIRM.

**Comment by kixelated** [comment]
> The proponents of publisher chosen IDs have cited both memory and performance as potential reasons for requiring this (please add other reasons if I missed them).
> 
> For memory it can save 8 bytes per subscription with confirmed ID. Is there any back-of-the-envelope math indicating this adds up to something significant? What's the target number of active subscriptions per-server (1M, 10M, 100M, 1B)?

Subscriber IDs are only going to use 1-2 bytes if they increment from 0 like QUIC stream IDs. This is beyond inconsequential and something is wrong if the size of `struct subscribe` matters. I would like to see a `MAX_SUBSCRIBES` message if it does matter to prevent any resource exh

Meanwhile hashed publisher IDs will use almost 8 bytes on average. This is inconsequential, 

Like Victor said in the meeting, any performance benefits depend on the QUIC library. If your QUIC library uses something like `iov` for writes, there's no performance impact. That's how Quinn works; and all QUIC streams share the underlying `Bytes`

Quinn works by

**Comment by kixelated** [comment]
Sorry misclick on my phone

**Comment by kixelated** [comment]
> The proponents of publisher chosen IDs have cited both memory and performance as potential reasons for requiring this (please add other reasons if I missed them).
> 
> For memory it can save 8 bytes per subscription with confirmed ID. Is there any back-of-the-envelope math indicating this adds up to something significant? What's the target number of active subscriptions per-server (1M, 10M, 100M, 1B)?

Subscriber IDs are only going to use 1-2 bytes if they increment from 0 like QUIC stream IDs. This is beyond inconsequential. The only time it might matter is if there's an attack involving spurious subscriptions, in which case I would like to see `MAX_SUBSCRIBES` regardless.

Note that the publisher could choose to cache the object using a 8 byte VarInt. It could then use memcopy to set the ID on a per subscription basis. This will slightly increase the OBJECT overhead on the wire, but so will using a random hash for the subscription ID (nearly 8 bytes on average).

And like Victor said in the meeting, this is a performance non-issue if your QUIC library supports something like `iov` for writes. Each subscription would share a pointer to the underlying OBJECT payload. The only difference is that the OBJECT header would be different per subscription, but again you could use iov to avoid any copies.

That's actually supported by Quinn and I use it on moq-rs. QUIC streams share a `Bytes` struct which is recounted under the hood.

If your QUIC library doesn't support `iov`, well then you're making a copy per QUIC stream anyway, even if the bytes are identical. This publisher side optimization is ONLY relevant if your QUIC library accepts a single pointer for the entire stream payload.

**Comment by kixelated** [comment]
I started migrating to the current draft but I gave up.

With #258 every subscription is stored in a lookup table. When I receive an `OBJECT` or `SUBSCRIBE_*` message, I use a lookup table to fetch the current state. This is possible because `SUBSCRIBE` contains the `track_id`:

```rust
struct Publisher {
  // maps from track_id to subscription
  subscriptions: HashMap<VarInt, track::Publisher>,
}
```

I started implementing the current draft where `SUBSCRIBE_OK` contains `track_id`, and this is what I think is needed:
```rust
struct Publisher {
  // maps from (namespace, name) to subscription
  subscriptions: HashMap<(String, String), track::Publisher>,

  // maps from track_id to (namespace, name) after receiving SUBSCRIBE_OK
  aliases: HashMap<VarInt, Deferred<(String, String)>>,
}
```

`Deferred` doesn't exist in Rust but my point is that it's an asynchronous lookup table.  I need to check if there's an entry in `aliases` or create one, which will be resolved when/if `SUBSCRIBE_OK` arrives. My `OBJECT` decoder will block until `SUBSCRIBE_OK` arrives, which might actually never happen if an attacker sends OBJECTs with random `track_id`s.

It's just a pain in the butt. The performance arguments are moot if you need to perform a 2nd lookup for each OBJECT. I think you can avoid it with separate lookup tables, but that adds more complexity.

#305 replaces `(String, String)` with `VarInt` but it's otherwise the same problem.

**Comment by ianswett** [comment]
Conclusion of the session was to add a way to reject a SUBSCRIBE and provide a different track ID, and have a separate Track ID from Subscribe ID.

**Comment by afrind** [comment]
Draft-02 allows the publisher to request a different track alias.
